engine/access/ingestion2/results_forest_test.go

```diff
@@ -0,0 +1,1688 @@
+	receiptGen := s.g.ExecutionReceipts()
+
+	forest := s.createForest()
+
+	// NewPipeline will be called at least once for each result. it may be called more than once
+	// depending on timing.
+	for _, result := range s.results {
+		pipeline := opsyncmock.NewPipeline(s.T())
+		pipeline.On("GetState").Return(optimistic_sync.StatePending).Maybe()
+		s.pipelineFactory.
+			On("NewPipeline", result).
+			Return(pipeline)
+	}
+
+	receiptsPerResult := 100
+
+	// mark the first 2 blocks as finalized so we can use different result statuses
+	for _, block := range s.blocks[:2] {
+		err := forest.OnBlockFinalized(block)
+		s.Require().NoError(err)
+	}
+
+	wg := sync.WaitGroup{}
+	for i := range len(s.results) * receiptsPerResult {
+		wg.Go(func() {
+			index := i % len(s.results)
+			status := ingestion2.ResultForCertifiedBlock
+			if index < 2 {
+				status = ingestion2.ResultForFinalizedBlock
+			}
+			receipt := receiptGen.Fixture(receiptGen.WithExecutionResult(*s.results[index]))
+			added, err := forest.AddReceipt(receipt, status)
+			s.Require().NoError(err)
+			s.True(added)
+		})
+	}
+	wg.Wait()
+
+	// verify all of the results were added and have the correct number of receipts
+	for i, result := range s.results {
+		container, found := forest.GetContainer(result.ID())
+		s.True(found)
+		s.Equal(result.ID(), container.ResultID())
+		s.Equal(uint(receiptsPerResult), container.Size())
+		if i < 2 {
+			s.Equal(ingestion2.ResultForFinalizedBlock, container.ResultStatus())
+		} else {
+			s.Equal(ingestion2.ResultForCertifiedBlock, container.ResultStatus())
+		}
+	}
+}
+
+func (s *ResultsForestSuite) TestHasReceipt() {
+	receiptGen := s.g.ExecutionReceipts()
+
+	forest := s.createForest()
+
+	result := s.results[0]
+	receipt := receiptGen.Fixture(receiptGen.WithExecutionResult(*result))
+
+	// receipt should not be in the forest initially
+	has := forest.HasReceipt(receipt)
+	s.False(has)
+
+	_ = s.createPipeline(result)
+	added, err := forest.AddReceipt(receipt, ingestion2.ResultForCertifiedBlock)
+	s.Require().NoError(err)
+	s.True(added)
+
+	// now it should be in the forest
+	has = forest.HasReceipt(receipt)
+	s.True(has)
+}
+
+// TestOnBlockFinalized tests the OnBlockFinalized handler by creating a forest with execution and
+// consensus forks, then verifying the forks are abandoned as expected.
+//
+// This test works with 4 forks:
+// - main chain
+// - execution fork (2 results with same blocks)
+// - 2 consensus forks (different blocks and results)
+//
+// It them walks through different scenarios of finalizing the blocks to test the forest's behavior.
+func (s *ResultsForestSuite) TestOnBlockFinalized() {
+	lastSealedResult := s.results[0]
+	mainResults := s.results[1:]
+	execFork := s.executionFork(s.blocks[1:], lastSealedResult.ID())
+	consensusFork1 := s.consensusFork(2, s.blocks[0].ToHeader(), lastSealedResult.ID())
+	consensusFork2 := s.consensusFork(2, s.blocks[0].ToHeader(), lastSealedResult.ID())
```

Comment:
[optional] I suspect we are constructing two certified blocks here for at least some of the views. This cannot happen in our consensus. I am a big fan of testing with realistic scenarios.